src/utils/src/Threadpool.c

```diff
@@ -319,15 +321,18 @@ STATUS threadpoolFree(PThreadpool pThreadpool)
                 finished = TRUE;
                 break;
             }
-            retStatus = stackQueueIteratorGetItem(iterator, &item);
+
+            CHK_STATUS(stackQueueIteratorGetItem(iterator, &data));
+            item = (PThreadData) data;
+            CHK(item != NULL, STATUS_INTERNAL_ERROR);
```

Comment:
If the iterator is still valid, this means it may not be the last item in the thread list. We would need to handle all other items before exiting this loop, or the corresponding threads will not terminate gracefully.

Reply:
The paradigm currently employed is the following sequence of calls in a loop:
```
stackQueueGetIterator(pThreadpool->threadList, &iterator);
stackQueueIteratorGetItem(iterator, &data)
stackQueueRemoveItem(pThreadpool->threadList, data)
```

In order for the iterator to actually iterate we need to remove the item.  However the removal happens under the `item->dataMutex` lock... but this is the very thing that will crash if `item` is `NULL`.  Are you saying that it's actually possible for there to be `PThreadData` on the stack/queue with value of `NULL` and they need to be discarded and we need to keep iterating? 

If I understood correctly then I suppose that is possible by checking if item is NULL instead of returning a status null error we just remove the NULL item and keep going, without attempting to acquire the mutex or setting the terminate flag (since there is no actual item object available).  

Ideally it should not be possible to have `NULL` items on this stackQueue, but I'm not sure if that was part of your design for the threadpool?

Reply:
Looking at this block it looks like it is not possible to ever have a NULL (0) item as a data in this queue:

```
    data->dataMutex = MUTEX_CREATE(FALSE);
    data->pThreadpool = pThreadpool;
    ATOMIC_STORE_BOOL(&data->terminate, FALSE);

    CHK_STATUS(stackQueueEnqueue(pThreadpool->threadList, (UINT64) data));
```

But if we did have `NULL` somehow that would imply some of corruption and in that case returning an internal failure seems appropriate.  Alternatively we could "remove" the `NULL` items and keep going as suggested earlier?    

The only way the `stackQueueIteratorGetItem(iterator, &data)` can error out is if the iterator itself is `NULL` in which case it is not valid and terminating is the correct thing to do.  Or if `&data` is `NULL` which is impossible.

Reply:
It shouldn't be possible for item to be NULL unless something has gone wrong with the stackqueue. However since we're checking for it we should fully handle the use case.

If for whatever reason the item is null, but the iterator is still valid from the earlier check then we need to remove the item OR call `stackQueueIteratorNext()` until we've fully gone through the list.

It's not important that the item mutex be locked when removing it from the stack queue if the item itself is NULL.

Reply:
Sure.  Previously before adding the check it would have crashed, now I've modified it to remove the NULL item and keep iterating. I added a WARN level log message if it is actually ever NULL since as you said that should not be possible.